### PR TITLE
overlord/snapstate, store: set a header when auto-refreshing

### PR DESCRIPTION
**Pull request by chipaca**
With this change, overlord/snapstate always passes in a flag to the
store when it's auto-refreshing, and the store uses this to set a
`Snap-Refresh-Reason: scheduled` header.

In future we can also use this flag to select a different retry
strategy for auto-refreshes, for example.

### CODE REVIEW

File: overlord/snapstate/handlers.go [hunk 1]
```diff
@@ -444,7 +444,7 @@ func (m *SnapManager) doDownloadSnap(t *state.Task, tomb *tomb.Tomb) error {
 	meter := NewTaskProgressAdapterUnlocked(t)
 	targetFn := snapsup.MountFile()
 
-	dlOpts := &store.DownloadOptions{}
+	dlOpts := &store.DownloadOptions{IsAutoRefresh: snapsup.IsAutoRefresh}
```
Comment (pedronis) [inline overlord/snapstate/handlers.go]: not really related to this but the if below is a bit strange and now is stranger, I would prefer things turned around:

    if !snapsup.IsAutoRefresh {
        rate = 0
     }
     dlOpts := &store.DownloadOptions{IsAutoRefresh: snapsup.IsAutoRefresh, RateLimit: rate}

File: overlord/snapstate/snapstate.go [hunk 2]
```diff
@@ -774,7 +774,8 @@ func updateManyFiltered(ctx context.Context, st *state.State, names []string, us
 		return nil, nil, err
 	}
 
-	updates, stateByInstanceName, ignoreValidation, err := refreshCandidates(ctx, st, names, user, nil)
+	refreshFlags := &store.RefreshOptions{IsAutoRefresh: flags.IsAutoRefresh}
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go]: this should be refreshOpts, not Flags.

Is this bit of change tested?  not obvious looking around

File: store/store.go [hunk 3]
```diff
@@ -1411,7 +1413,7 @@ func (s *Store) Download(ctx context.Context, name string, targetPath string, do
 	return s.cacher.Put(downloadInfo.Sha3_384, targetPath)
 }
 
-func reqOptions(storeURL *url.URL, cdnHeader string) *requestOptions {
+func reqOptions(storeURL *url.URL, cdnHeader string, opts *DownloadOptions) *requestOptions {
```
Comment (pedronis) [inline store/store.go]: again not related to this, but the change make clear that this function should mention GET somehow in its name

Comment (chipaca) [reply store/store.go]: hah, when I saw it I thought "hmm, this should probably say Download somewhere in its name"

File: overlord/snapstate/snapstate.go (hunk 2 above)
Comment (chipaca) [reply overlord/snapstate/snapstate.go]: Yes, via SnapAction on the autoRefreshStore. Not super obvious

File: overlord/snapstate/autorefresh_test.go [hunk 4]
```diff
@@ -49,6 +49,10 @@ type autoRefreshStore struct {
 }
 
 func (r *autoRefreshStore) SnapAction(ctx context.Context, currentSnaps []*store.CurrentSnap, actions []*store.SnapAction, user *auth.UserState, opts *store.RefreshOptions) ([]*snap.Info, error) {
+	if !opts.IsAutoRefresh {
+		panic("AutoRefresh snap action did not set IsAutoRefresh flag")
```
Comment (stolowski) [inline overlord/snapstate/autorefresh_test.go]: Slightly surprised to see panic here, I'd expect an internal error (apparently this is the approach taken in the existing checks below).

Comment (chipaca) [reply overlord/snapstate/autorefresh_test.go]: surprising *and* unhelpful, so I'll probably be changing it in a separate PR.

As it stands, if any of these checks fail, you get a very unhelpful panic about the state lock, not about the actual error.

**Pull request merged**